protocol/messaging/entrypoint-contracts.md

```diff
@@ -0,0 +1,134 @@
+## Summary
+
+This design document introduces `Entrypoint` contracts as a new primitive that allows anyone to add custom logic on top of the `L2ToL2CrossDomainMessenger`. It generalizes the `L2ToL2CrossDomainMessenger` design and unlocks other interop primitives such as message batching and expiring. To do so, it adds a parameter in the `L2ToL2CrossDomainMessenger` that binds the relaying to a particular address, where custom logic can live.
+
+## Problem Statement + Context
+
+Building functionality on top of the `L2ToL2CrossDomainMessenger` is not simple. For example, to allow message recovery, the contract would need to track expired/failed messages to prevent someone from relaying them after recovery. This would require adding code on the `L2ToL2CrossDomainMessenger`. If another feature were required, we would need to add more code, which doesn’t scale.
+
+The `Entrypoint` contract primitive solves this issue and unlocks new powerful cross-chain actions. The integration with `Entrypoint` contracts is simple, and the code difference is small.
+
+## `Entrypoint` Contracts
+
+### Main idea
+
+Anyone can deploy `Entrypoint` contracts with any logic. Users can indicate, explicitly or through a contract, which `Entrypoint` contract can process their messages in the destination chain. `relayMessage` will check if the message has an `Entrypoint` associated, and if it does, it will validate that this contract is the `msg.sender`; otherwise, it will revert.
+
+The key modification is the `msg.sender` binding feature on the `relayMessage` call. If `msg.sender` has logic, the design is equivalent to binding `relayMessage` to include a pre-hook with that logic.
+
+Notice that a pre/post-hook design is already possible with the current design.
+
+### Summary of required changes
+
+The following changes are required:
```

Comment:
I believe that we agreed to overload `sendMessage`? Would that be possible with the diff required to `relayMessage`

Reply:
Good point. This would be possible, but we would need to prevent `relayMessage` to be successful. 
- One way could be to keep the selector check [here](https://github.com/ethereum-optimism/optimism/blob/bc9b6cd588588c9c4167c926a1782c658e5df921/packages/contracts-bedrock/src/L2/L2ToL2CrossDomainMessenger.sol#L225), which I think is going to be deprecated as per this [comment](https://github.com/ethereum-optimism/optimism/blob/bc9b6cd588588c9c4167c926a1782c658e5df921/packages/contracts-bedrock/src/L2/L2ToL2CrossDomainMessenger.sol#L60).
- The other is to encode the `entrypoint` first in the list, so that decoding succeeds, but `relayMessage` will fail on [this check](https://github.com/ethereum-optimism/optimism/blob/bc9b6cd588588c9c4167c926a1782c658e5df921/packages/contracts-bedrock/src/L2/L2ToL2CrossDomainMessenger.sol#L169).

I feel like the selector is more explicit, as it will throw the correct error. Do you have any context on the removal of it?

Reply:
@tynes an alternative here is to rewrite `sendMessage` with the `entrypoint` parameter, an do not revert in `relayMessage` if entrypoint is `address(0)`.